docs/design/data_safety.md

```diff
@@ -0,0 +1,182 @@
+
+[^2]: *"Expensive" refers to both code size and performance.*
+
+[^3]: *If an attacker is able to mutate ICU4X data sources, they could construct data to produce a result they desire, regardless of whether unexpected operations end in GIGO. However, panicky algorithms could enable attackers to perform denial-of-service attacks.*
+
+Additional points:
+
+1. As a rule of thumb, validation that requires a single linear pass over the data with no memory allocations is fine, similar to UTF-8 validation, but validation that requires memory allocations, runs in superlinear time, or pulls in large amounts of code or dependencies is discouraged.
+2. If performing binary search on a vector from data, it is not necessary to validate that the vector is sorted. The binary search will fail to find elements in an unsorted vector, but it won't panic, so this is consistent with GIGO.
+3. The best data structs are those that don't need to be validated. See the below example on how to write data structs without internal invariants.
+
+## Example
+
+Consider the struct
+
+```rust
+#[derive(serde::Deserialize)]
+pub struct MonthNamesBad {
+    // Invariant 1: the first element in the vector is the default value
+    // Invariant 2: the length of the vector is the same as month_count
+    // Invariant 3: month_count is at least 1
+    month_names: Vec<String>,
+    month_count: usize,
+}
+
+impl MonthNamesBad {
+    pub fn get_first_month_name(&self) -> &str {
+        &self.month_names[0]
+    }
+    /// Returns None only if `idx` is greater than the month count.
+    pub fn get_month_name_at_index(&self, idx: usize) -> Option<&str> {
+        if idx >= self.month_count {
+            return None;
+        }
+        Some(&self.month_names[idx])
+    }
+    pub fn get_month_count(&self) -> usize {
+        self.month_count
+    }
+}
+```
+
+This struct has 2 invariants, and each of the functions could panic if those invariants are not upheld.
+
+Below are 3 approaches to resolve this issue that are consistent with the ICU4X style.
+
+*Editor's note: the examples in this document are tested along with the tutorials.*
+
+### Solution 1: Garbage In, Garbage Out
+
+Change the functions to return default fallback values if the data is not in the expected form. This code also consolidates `month_count` and `month_names` into the same field, a change reflected in all of the proposed solutions.
+
+```rust
+#[derive(serde::Deserialize)]
+pub struct MonthNamesGIGO {
+    // WEAK Invariant: the first element in the vector is the default value
+    month_names: Vec<String>,
+}
+
+impl MonthNamesGIGO {
+    pub fn get_first_month_name(&self) -> &str {
+        match self.month_names.get(0) {
+            Some(v) => v,
+            None => {
+                debug_assert!(false, "month_names is empty");
+                ""
```

Comment:
I'd call out that here you are returning a "garbage" response `""` in case the data is not upholding to invariants.